### ISSUE TITLE
Use out-of-band buffers for Python buffer protocol supporting objects

### ISSUE THREAD

**Issue by jakirkham** [cloudpipe/cloudpickle#365]
For objects supporting [Python's buffer protocol]( https://docs.python.org/3.8/c-api/buffer.html ) (like `memoryview`), it would be great to have a path where we use [Pickle's protocol 5 for serialization of out-of-band buffers]( https://docs.python.org/3.8/library/pickle.html#out-of-band-buffers ). This should avoid some copies when pickling/unpickling and allow more efficient transmission.

**Comment by pierreglaser** [comment]
Sure. Is there any reason for why `CPython` does not have a `memoryview` reducer?

**Comment by jakirkham** [comment]
That's a good question. I don't know. @pitrou, do you know? 🙂

**Comment by pitrou** [comment]
Because it's not obvious how to serialize a `memoryview`.

**Comment by jakirkham** [comment]
Meaning we don't know what memory to use for backing it?

**Comment by pitrou** [comment]
Yes, exactly. One way of punting the question is to decide to deserialize a memoryview by recreating a bytes object with similar content. But what if your `memoryview` was backed by e.g. a mmap'ed file?

**Comment by jakirkham** [comment]
Yeah that makes sense. I think in cloudpickle we have [chosen to just use `bytearray`]( https://github.com/cloudpipe/cloudpickle/blob/0dd91b48ee4f89074c95238c7477775fd6105893/cloudpickle/cloudpickle.py#L490-L491 ), but it makes sense this would be ambiguous generally and may not make sense outside of cloudpickle.

What about other objects like `bytes`, `bytearray`, `array`, etc. where we do know how to allocate for them? Does it make sense for CPython to support out-of-band buffer serialization for them or are there some issues there as well?

**Comment by pitrou** [comment]
It would certainly make sense indeed. But it's not necessarily trivial and someone has to do the work ;-)

(note that a memoryview may have been sliced or casted... of course, you could choose to forbid those cases, but you would still have to detect them)

**Comment by jakirkham** [comment]
What challenges do you envision in implementing out-of-band serialization for those other objects?

(yeah that's fair. there is a lot the current implementation brushes under the rug ;)

**Comment by jakirkham** [comment]
FWIW opened upstream issue ( https://bugs.python.org/issue40718 ) on serializing builtin `bytes`-like types.

**Comment by jakirkham** [comment]
The one trick for performing out-of-band pickling of `memoryview`s is we would need access to the protocol version (as is the case with [`__reduce_ex__`]( https://docs.python.org/3.8/library/pickle.html#object.__reduce_ex__ ). A function won't work as we can't pass more arguments to the function when using the `dispatch_table`. Maybe we can workaround this constraint by using a method and accessing the protocol version via class state. Thoughts?
